### DISCUSSION TITLE
How to dynamically schedule the workflow via cron

### DISCUSSION THREAD

**Discussion by NiveditaGalwade** [elsa-workflows/elsa-core#5279] [Q&A]
How to dynamically schedule the workflow via cron

**Comment by glime-ai** [comment]
To dynamically schedule a workflow using a cron expression in Elsa Workflows, you can use the Timer trigger activity, which allows you to specify a cron expression that determines when the workflow should be triggered.

Here are the steps to schedule a workflow with a cron expression:

1. **Create a Timer Trigger Activity**: In your workflow definition, add a Timer trigger activity where you can set the cron expression.

2. **Set the Cron Expression**: Configure the Timer trigger activity with the cron expression that represents your desired schedule. For example, `0 * * * *` would trigger the workflow at the top of every hour.

3. **Handle the Trigger**: Ensure that your workflow is designed to handle the trigger correctly. This means that when the Timer trigger fires based on the cron schedule, the workflow should execute the subsequent activities as intended.

4. **Publish the Workflow**: After setting up the Timer trigger with the cron expression, publish the workflow so that it becomes active.

5. **Monitor the Workflow**: Once published, the workflow should be triggered automatically according to the schedule specified by the cron expression.

Please note that if you encounter issues with the workflow not triggering as expected, you may need to check the logs or the Elsa Dashboard for any errors or misconfigurations.

For more detailed information and examples, you can refer to the Elsa Workflows documentation, specifically the sections on triggers and invoking workflows.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by RescueRangers** [reply]
Sorry to dig up a  year old question but it's the first relevant thing that shows up for me when I search for elsa 3 cron.

Ad 1. Timer trigger does not have an option for cron expression, I am assuming it's about cron trigger activity
Ad 2. The example cron expression returns an error: "The given cron expression has an invalid format. Months: Unexpected character '\0'."

So what is the correct cron expression format that elsa is expecting?

**Comment by sfmskywalker** [comment]
For Elsa 3, use the **Cron** activity as the workflow trigger; the earlier reply's reference to a generic Timer trigger is not the precise Elsa 3 guidance.

Register scheduling with `elsa.UseScheduling()`, add a Cron activity at the start of the workflow, set its `CronExpression`, and publish the definition. The Cron activity accepts an `Input<string>`, so the expression can be a literal or an Elsa expression. For example, the current component test uses `* * * * * *` for every second:

```csharp
builder.Root = new Sequence
{
    Activities =
    {
        Elsa.Scheduling.Activities.Cron.FromCronExpression("* * * * * *"),
        new MyCustomActivity()
    }
};
```

For a user-editable schedule, persist the chosen expression in the workflow definition and publish the revised definition so Elsa can re-index its trigger. If the schedule needs to vary per user, model that as separate workflow definitions (or a definition you update through the management API), rather than changing an already-running workflow instance.

Relevant source:

- https://github.com/elsa-workflows/elsa-core/blob/main/src/modules/Elsa.Scheduling/Activities/Cron.cs
- https://github.com/elsa-workflows/elsa-core/blob/main/test/component/Elsa.Workflows.ComponentTests/Scenarios/Activities/Timers/Cron/Workflows/CronWorkflow.cs

Closing this as answered. If you meant runtime rescheduling of one existing instance rather than editing a workflow definition, please open a new discussion with the Elsa version and desired lifecycle.
